Build a storage reader/writer to write checkpoints in HF format

ankitageorge opened the pull request:
Title - we want to write checkpoints in HF format with DCP, this diff allows this for the non-distributed use case. 
Copy of [D68444967](https://www.internalfb.com/diff/D68444967) (https://github.com/pytorch/pytorch/pull/146352). That diff got reverted because of lint errors. The lint error was due to having imports of uninstalled libraries. This was on purpose because we don't want to install safetensors and huggingface, this new diff explicitly ignores this lint so that we don't have the error.




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

Skylion007 commented on torch/distributed/checkpoint/_hf_storage.py:
```diff
@@ -0,0 +1,201 @@
+            path: hf directory where the checkpoint will be read from. Should begin with hf://.
+            token: The token to use to authenticate with huggingface hub.
+        """
+        import fsspec  # type: ignore[import-untyped]
```
Like wise, any reason for the lazy import? Is it really not imported by FsspecReader?

ankitageorge commented on torch/distributed/checkpoint/_fsspec_filesystem.py:
```diff
@@ -8,7 +8,11 @@
 from pathlib import Path
 from typing import Optional, TYPE_CHECKING, Union
 
-from fsspec.core import url_to_fs
```
A lot of internal tests were failing without making it optional. With the import of torch.distributed.checkpoint, the HfFileSystem imports are becoming required, but I guess many of the use cases don't have fsspec installed.

saumishr replied:
For internal tests, we can rather have a dependency on fsspec. Its available as a third-party library and avoid making it an optional dependency here.

ankitageorge committed:
Build a storage reader/writer to write checkpoints in HF format (#147622)

Summary:

Title - we want to write checkpoints in HF format with DCP, this diff allows this for the non-distributed use case.
Copy of D68444967 (https://github.com/pytorch/pytorch/pull/146352). That diff got reverted because of lint errors. The lint error was due to having imports of uninstalled libraries. This was on purpose because we don't want to install safetensors and huggingface, this new diff explicitly ignores this lint so that we don't have the error.

Test Plan:
buck2 test 'fbcodemode/dev-nosan' fbcodecaffe2/test/distributed/checkpoint:test_hf_torchtune_storage
N6596613 --> able to save and load tensor in hf format

Differential Revision: D69984656
